tests/test_cli.py

```diff
@@ -0,0 +1,140 @@
+     'del_range': ['400-404', '500'], 'del_error': {400, 401, 402, 403, 404, 500}},
+])
+@pytest.mark.parametrize('value_params', [
+    {'add_tags': ['foo,bar', 'baz'], 'tags_fetch': False, 'tags_in': ',bar baz,foo,', 'title': ''},
+    {'tag': ['foo', 'bar,baz'], 'tags_fetch': False, 'tags_in': ',baz,foo bar,', 'title': 'Custom Title'},
+    {'add_tags': ['+', 'foo', 'bar', 'baz'], 'tags_in': ',foo bar baz,', 'comment': ''},
+    {'tag': ['+', 'foo,bar,baz'], 'tags_in': ',bar,baz,foo,', 'comment': 'Custom Description'},
+    {'add_tags': ['-', 'foo', 'baz', 'baz'], 'tags_except': ',foo baz baz,', 'immutable': False},
+    {'tag': ['-', 'foo,', ',baz,', ',baz'], 'tags_except': ',baz,foo,', 'immutable': True},
+    {'add_tags': ['foo,baz,bar'], 'tag': ['baz,qux'],
+     'tags_fetch': False, 'tags_in': ',bar,baz,foo,qux,'},
+    {'add_tags': ['+', 'foo,baz,bar'], 'tag': ['baz,', 'qux,'],
+     'tags_fetch': False, 'tags_in': ',bar,baz,foo,qux,'},
+    {'add_tags': ['foo,baz,', 'bar,'], 'tag': ['+', 'baz,qux'],
+     'tags_fetch': False, 'tags_in': ',bar,baz,foo,qux,'},
+    {'add_tags': ['-', 'foo,baz,bar,'], 'tag': ['baz,', 'qux,'],
+     'tags_fetch': False, 'tags_in': ',baz,qux,', 'tags_except': ',bar,baz,foo,'},
+    {'add_tags': ['foo,baz,', 'bar,'], 'tag': ['-', 'baz,qux'],
+     'tags_fetch': False, 'tags_in': ',bar,baz,foo,', 'tags_except': ',baz,qux,'},
+    {'add_tags': ['-', 'foo,baz,bar,'], 'tag': ['-', 'baz,', 'qux,'], 'tags_except': ',bar,baz,foo,qux,'},
+    {'add_tags': ['-', 'foo,baz,', 'bar,'], 'tag': ['+', 'baz,', 'qux,'],
+     'tags_in': ',baz,qux,', 'tags_except': ',bar,baz,foo,'},
+    {'add_tags': ['+', 'foo,baz,', 'bar,'], 'tag': ['-', 'baz,', 'qux,'],
+     'tags_in': ',bar,baz,foo,', 'tags_except': ',baz,qux,'},
+])
+def test_add(stdin, bdb, prompt, value_params, fetch_params):
+    _test_add(bdb, prompt, **value_params, **fetch_params)
+
+def _test_add(bdb, prompt, *, add_tags=[], tag=[], tags_fetch=True, tags_in=None, tags_except=None,
+              title=None, comment=None, immutable=None, offline=False, url_redirect=False,
+              tag_redirect=False, tag_error=False, del_range=None, del_error=None):
+    argv = ['--add', (url := 'https://example.com/')] + add_tags
+    if tag:
+        argv += ['--tag'] + tag
+    if title is not None:
+        argv += ['--title', title]
+    if comment is not None:
+        argv += ['--comment', comment]
+    if immutable is not None:
+        argv += ['--immutable', str(int(immutable))]
+    if offline:
+        argv += ['--offline']
+    if url_redirect:
+        argv += ['--url-redirect']
+    if tag_redirect:
+        argv += ['--tag-redirect'] + ([] if isinstance(tag_redirect, bool) else [tag_redirect])
+    if tag_error:
+        argv += ['--tag-error'] + ([] if isinstance(tag_error, bool) else [tag_error])
+    if del_error:
+        argv += ['--del-error'] + del_range
+    with pytest.raises(SystemExit):
+        buku.main(argv)
```

Comment:
Yanno, maybe we should close the DB _without_ calling `sys.exit()`, at least from within `main()` ­– that would make it a better fit for external use :sweat_smile: